Classes/common/touchdb/TD_DatabaseManager.m

```diff
@@ -117,39 +117,60 @@ - (NSString*)pathForName:(NSString*)name
     return [_dir stringByAppendingPathComponent:[name stringByAppendingPathExtension:kDBExtension]];
 }
 
-- (TD_Database*)databaseNamed:(NSString*)name create:(BOOL)create
+- (TD_Database*)databaseNamed:(NSString*)name
+            withEncryptionKey:(id<CDTEncryptionKey>)encryptionKey
+                       create:(BOOL)create
 {
-    if (_options.readOnly) create = NO;
+    if (_options.readOnly) {
+        create = NO;
+    }
+
     TD_Database* db = _databases[name];
     if (!db) {
         NSString* path = [self pathForName:name];
-        if (!path) return nil;
-        db = [[TD_Database alloc] initWithPath:path];
+        if (!path) {
+            return nil;
+        }
+
+        db = [[TD_Database alloc] initWithPath:path encryptionKey:encryptionKey];
         db.readOnly = _options.readOnly;
         if (!create && !db.exists) {
             return nil;
         }
+
         db.name = name;
         _databases[name] = db;
     }
+
     return db;
 }
 
-- (TD_Database*)databaseNamed:(NSString*)name { return [self databaseNamed:name create:YES]; }
+- (TD_Database*)databaseNamed:(NSString*)name withEncryptionKey:(id<CDTEncryptionKey>)encryptionKey
+{
+    return [self databaseNamed:name withEncryptionKey:encryptionKey create:YES];
+}
 
 - (TD_Database*)existingDatabaseNamed:(NSString*)name
+                    withEncryptionKey:(id<CDTEncryptionKey>)encryptionKey
 {
-    TD_Database* db = [self databaseNamed:name create:NO];
-    if (db && ![db open]) db = nil;
+    TD_Database* db = [self databaseNamed:name withEncryptionKey:encryptionKey create:NO];
+    if (db && ![db open]) {
+        db = nil;
+    }
+
     return db;
 }
 
-- (BOOL)deleteDatabaseNamed:(NSString*)name
+- (BOOL)deleteDatabaseNamed:(NSString*)name withEncryptionKey:(id<CDTEncryptionKey>)encryptionKey
 {
-    TD_Database* db = [self databaseNamed:name];
-    if (!db) return NO;
+    TD_Database* db = [self databaseNamed:name withEncryptionKey:encryptionKey];
```

Comment:
Before deleting a database, a notification is sent passing the database as a parameter. As far as I can tell, only `TDReplicator` is observing this notification.

But the key is required only if the database is not open yet. If the database is not open, no replicator can be attached to it. So I think I can simply check if the database is open and, if it is, I can delete it as I do now. In other case, to delete a database I only have to remove the database as well as the attachment from disk. I can re-factor some of the code in `TD_Database` and add a couple of class methods to do so.